packages/css-selector-parser/src/ast-tools/specificity.ts

```diff
@@ -0,0 +1,105 @@
+import type {
+  ImmutableSelectorNode,
+  ImmutablePseudoClass,
+  SelectorNode,
+} from "../ast-types";
+import { walk } from "./walk";
+
+type Specificity = [number, number, number, number];
+export function calcSpecificity(ast: ImmutableSelectorNode): Specificity {
+  const result: Specificity = [0, 0, 0, 0];
+  // ToDo: remove casting once immutable walk is supported
+  walk(ast as SelectorNode, (node) => {
+    switch (node.type) {
+      case `type`:
+      case `pseudo_element`:
+        result[3]++;
+        break;
+      case `class`:
+      case `attribute`:
+        result[2]++;
+        break;
+      case `pseudo_class`:
+        if (customPseudoClass[node.value]) {
+          customPseudoClass[node.value](node, result);
+          return walk.skipNested;
+        }
+        result[2]++;
+        break;
+      case `id`:
+        result[1]++;
+        break;
+    }
+    return node.type !== `selector` && node.type !== `compound_selector`
+      ? walk.skipNested
+      : undefined;
+  });
+  return result;
+}
+
+const customPseudoClass: Record<
+  string,
+  (node: ImmutablePseudoClass, result: Specificity) => void
+> = {
+  not: mostSpecificInnerSelector,
+  is: mostSpecificInnerSelector,
+  has: mostSpecificInnerSelector,
+  where: () => {
+    /* no specificity*/
+  },
+  "nth-child": pseudoClassPlusMostSpecificInnerSelector,
+  "nth-last-child": pseudoClassPlusMostSpecificInnerSelector,
+  "nth-of-type": pseudoClassPlusMostSpecificInnerSelector,
+  "nth-last-of-type": pseudoClassPlusMostSpecificInnerSelector,
+};
+
+function pseudoClassPlusMostSpecificInnerSelector(
+  node: ImmutablePseudoClass,
+  result: Specificity
+) {
+  result[2]++;
+  mostSpecificInnerSelector(node, result);
+}
+function mostSpecificInnerSelector(
+  node: ImmutablePseudoClass,
+  result: Specificity
+) {
+  if (node.nodes) {
```

Comment:
maybe if you add a length check here you will be able to remove the `highest!`.
also is there are nodes with length 0 you will not have highest set. maybe add a check if(highest) {...}

Reply:
highest here is the reference to the current highest collected specificity - it start `undefined`.

if one of these special pseudo-class comes in with no content they are (probably?) not valid an break the selector and shouldn't add specificity

Reply:
but in this case it will throw an error....cannot get i of undefined

Reply:
unless addSpecificity will be able to get undefined and handle it

Reply:
Good catch, I will fix this